### PR TITLE
Improve bus write read transactions

### PR DESCRIPTION
**Pull request by kasjer**
Change extends bus driver operations to allow faster transfer.

New driver function **write_read** allow for faster transaction implementation.

Existing function **bus_node_write_read_transact()** was using **write** that **read**.
It turns out that for short transfers (1 or 2 bytes written) several bytes read it is worth
to implement custom function that writes and read.

This code also adds first implementation of such function for spi_hal driver that can
benefit around 50% time reduction on single transfer.

### CODE REVIEW

File: hw/bus/drivers/spi_hal/src/spi_hal.c [hunk 1]
```diff
@@ -191,6 +191,66 @@ bus_spi_write(struct bus_dev *bdev, struct bus_node *bnode, const uint8_t *buf,
...
+
+    /* XXX update HAL to accept const instead */
+
+#if MYNEWT_VAL(SPI_HAL_USE_NOBLOCK)
+    if (wlength + rlength <= sizeof(buf)) {
+        memcpy(buf, wbuf, wlength);
+        memset(buf + wlength, 0, rlength);
+        rc = hal_spi_txrx_noblock(dev->spi_dev.cfg.spi_num, buf, buf, wlength + rlength);
+        if (rc == 0) {
+            os_sem_pend(&dev->sem, OS_TIMEOUT_NEVER);
```
Comment (vrahane) [inline hw/bus/drivers/spi_hal/src/spi_hal.c:218]: This does need a specific timeout value. same with the one below.

**Comment by vrahane** [comment]
I do like the idea and I think this is needed for sure. As per above dummy reads/writes are handled by the underlying api which is great.

File: hw/bus/drivers/spi_hal/src/spi_hal.c (hunk 1 above)
Comment (kasjer) [reply hw/bus/drivers/spi_hal/src/spi_hal.c:218]: > This does need a specific timeout value. Same with the ones below.
Structure of the block was put together from write and read that use timeout like that.
While I'm not against changing this I will keep it like that in this PR.
If anyone feels that it should be changed, all places could be applied in separate PR(s).

**Review by vrahane — APPROVED** [review]
LGTM.

**Pull request merged**